File: examples/go/apps/minipi/agent.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// AgentActor — OODA-loop agent with durable execution trace capture.
//
// Demonstrates the harness layer: loop control, tool calling, token budget,
// crash recovery via DurabilityFacet, and execution trace export via
// ExecutionTraceFacet.
//
// Harness = agent - model (everything except the LLM).
package main

import (
	"encoding/json"
	"fmt"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

const (
	agentMaxIter    = 10
	agentTokenBudget = 4096
)

// AgentActor runs an OODA loop: Observe → Orient → Decide → Act.
//
// Durable: DurabilityFacet journals every step. Crash at step 7?
// Restart brings back all prior steps from journal — no re-burned tokens.
//
// Execution Trace: ExecutionTraceFacet captures ordered step sequence for eval,
// writes trace:{id} and trace_index:{actor_id} to KV on completion.
type AgentActor struct {
	plexspaces.BaseActor
	ActorID         string `json:"actor_id"`
	Task            string `json:"task"`
	IterationsDone  int    `json:"iterations_done"`
	TotalToolCalls  int    `json:"total_tool_calls"`
	EvalRunID       string `json:"eval_run_id"`
	ScenarioID      string `json:"scenario_id"`
}

func NewAgentActor() plexspaces.Actor {
	a := &AgentActor{}
	a.SetSelf(a)
	return a
}

func (a *AgentActor) Init(configJSON string) string {
	var config struct {
		ActorID string            `json:"actor_id"`
		Args    map[string]string `json:"args"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	a.SetRuntimeMetadata(config.ActorID)
	a.ActorID = config.ActorID
	if v, ok := config.Args["eval_run_id"]; ok {
		a.EvalRunID = v
	}
	if v, ok := config.Args["scenario_id"]; ok {
		a.ScenarioID = v
	}
	if err := host.PG().Join("svc:agents"); err != nil {
		host.Warn(fmt.Sprintf("AgentActor: failed to join svc:agents: %v", err))
	}
	host.Info(fmt.Sprintf("AgentActor Init actor_id=%s eval_run=%s", config.ActorID, a.EvalRunID))
	return ""
}

func (a *AgentActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "run", "workflow_run":
		return a.run(p)
	case "resume":
		host.Info(fmt.Sprintf("AgentActor resumed: %v", p))
		return marshal(map[string]any{"status": "ok", "resumed": true})
	case "execution_trace":
		return a.queryExecutionTrace()
	case "status":
		return a.queryStatus()
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

// Run implements WorkflowActor — framework calls this for behavior_kind = "Workflow".
func (a *AgentActor) Run(payloadJSON string) string {
	return a.Handle("", "workflow_run", payloadJSON)
}

// Signal implements WorkflowActor — handles resume/cancel signals.
func (a *AgentActor) Signal(name, _ string) {
	host.Info(fmt.Sprintf("AgentActor signal: %s", name))
}

// Query implements WorkflowActor — returns trace or status without side effects.
func (a *AgentActor) Query(name, _ string) string {
	switch name {
	case "execution_trace":
		return a.queryExecutionTrace()
	default:
		return a.queryStatus()
	}
}

func (a *AgentActor) run(p map[string]any) string {
	task := stringVal(p, "task", "")
	if evalRunID := stringVal(p, "eval_run_id", ""); evalRunID != "" {
		a.EvalRunID = evalRunID
	}
	if scenarioID := stringVal(p, "scenario_id", ""); scenarioID != "" {
		a.ScenarioID = scenarioID
	}
	if task == "" {
		return marshal(map[string]any{"error": "task is required"})
	}
	a.Task = task

	host.Info(fmt.Sprintf("AgentActor starting task: %.80s", task))

	cfg := plexspaces.AgentConfig{
		MaxIterations: agentMaxIter,
		TokenBudget:   agentTokenBudget,
		EvalRunID:     a.EvalRunID,
		ScenarioID:    a.ScenarioID,
	}
	actorID := a.ActorID
	if actorID == "" {
		actorID = "agent-unknown"
	}
	loop := plexspaces.NewAgentLoop(actorID, cfg)

	for !loop.IterationLimitReached() {
		if loop.BudgetExceeded() {
			a.IncrCounter(host, "agent_budget_exceeded")
			traj := loop.FinalizeTrajectory("budget_exceeded", fmt.Sprintf("Token budget %d exceeded", agentTokenBudget))
			a.exportTrajectory(traj)
			trajJSON, _ := json.Marshal(traj)
			return marshal(map[string]any{
				"status":      "budget_exceeded",
				"trajectory":  json.RawMessage(trajJSON),
				"trajectory_id": traj.TrajectoryID,
			})
		}

		if loop.IsSuspended() {
			traj := loop.GetTrajectory()
			trajJSON, _ := json.Marshal(traj)
			return marshal(map[string]any{
				"status":     "suspended",
				"trajectory": json.RawMessage(trajJSON),
				"trajectory_id": traj.TrajectoryID,
			})
		}

		// OBSERVE: fetch context from memory and environment
		observations := a.doObserve(loop, task)

		// ORIENT: analyze observations with LLM
		plan := a.doOrient(loop, observations)

		// DECIDE: pick next action
		action := a.doDecide(loop, plan)

		if boolVal(action.(map[string]any), "done") {
			break
		}

		// Check for approval-required actions
		if boolVal(action.(map[string]any), "needs_approval") {
			toolName := stringVal(action.(map[string]any), "tool_name", "unknown")
			loop.Suspend("action_needs_approval:" + toolName)
			traj := loop.GetTrajectory()
			trajJSON, _ := json.Marshal(traj)
			return marshal(map[string]any{
				"status":     "suspended",
				"trajectory": json.RawMessage(trajJSON),
				"trajectory_id": traj.TrajectoryID,
			})
		}

		// ACT: execute the chosen tool
		a.doAct(loop, action.(map[string]any))
		a.TotalToolCalls++
		a.IterationsDone++

		a.IncrCounter(host, "agent_iterations")
		loop.IncrementIteration()
	}

	traj := loop.FinalizeTrajectory("completed", fmt.Sprintf("Completed %d iterations", a.IterationsDone))
	a.exportTrajectory(traj)
	a.IncrCounter(host, "agent_runs_completed")
	trajJSON, _ := json.Marshal(traj)
	return marshal(map[string]any{
		"status":      "success",
		"task":        task,
		"iterations":  a.IterationsDone,
		"trajectory":  json.RawMessage(trajJSON),
		"trajectory_id": traj.TrajectoryID,
		"step_count":  len(traj.Steps),
		"outcome":     traj.Outcome,
	})
}

func (a *AgentActor) doObserve(loop *plexspaces.AgentLoop, task string) any {
	memoryKey := "agent_memory:" + a.ActorID
	priorContext := map[string]any{}
	if raw, _ := host.KV().Get(memoryKey); raw != "" {
		var m map[string]any
		if err := json.Unmarshal([]byte(raw), &m); err == nil {
			priorContext = m
		}
	}
	observations := map[string]any{
		"task":          task,
		"prior_context": priorContext,
		"iteration":     a.IterationsDone,
	}
	return loop.Observe(observations)
}

func (a *AgentActor) doOrient(loop *plexspaces.AgentLoop, observations any) any {
	obs, _ := observations.(map[string]any)
	if obs == nil {
		obs = map[string]any{}
	}
	llmID, err := registryFirst("llm_gateway", "svc:llm_gateway", "completion")

	var plan map[string]any
	if err != nil || llmID == "" {
		plan = map[string]any{
			"analysis":  fmt.Sprintf("Processing task: %v", obs["task"]),
			"next_tool": "calculator",
			"arguments": map[string]any{"expression": "1+1"},
			"done":      false,
		}
	} else {
		messages := []any{
			map[string]any{"role": "system", "content": "You are a helpful agent. Analyze the task and decide what to do next."},
			map[string]any{"role": "user", "content": fmt.Sprintf("Task: %v\nIteration: %v", obs["task"], obs["iteration"])},
		}
		resp, askErr := askActor(llmID, "completion", map[string]any{
			"op":       "completion",
			"messages": messages,
		}, 10000)
		if askErr != nil || len(resp) == 0 {
			plan = map[string]any{"done": true, "result": "LLM unavailable"}
		} else {
			response, _ := resp["response"].(map[string]any)
			if response == nil {
				response = map[string]any{}
			}
			stopReason := stringVal(response, "stop_reason", "end_turn")
			toolCalls := sliceVal(response, "tool_calls")
			plan = map[string]any{
				"analysis":      stringVal(response, "content", ""),
				"next_tool":     "calculator",
				"arguments":     map[string]any{},
				"input_tokens":  intVal(resp, "input_tokens", 0),
				"output_tokens": intVal(resp, "output_tokens", 0),
				"model":         stringVal(resp, "model", ""),
				"done":          stopReason == "end_turn" && len(toolCalls) == 0,
			}
		}
	}
	return loop.Orient(plan)
}

func (a *AgentActor) doDecide(loop *plexspaces.AgentLoop, plan any) any {
	planMap, _ := plan.(map[string]any)
	if planMap == nil {
		planMap = map[string]any{}
	}
	toolName := stringVal(planMap, "next_tool", "calculator")
	args, _ := planMap["arguments"].(map[string]any)
	if args == nil {
		args = map[string]any{}
	}
	action := map[string]any{
		"tool_name":      toolName,
		"arguments":      args,
		"done":           boolVal(planMap, "done"),
		"needs_approval": boolVal(planMap, "needs_approval"),
	}
	return loop.Decide(action)
}

func (a *AgentActor) doAct(loop *plexspaces.AgentLoop, action map[string]any) any {
	toolName := stringVal(action, "tool_name", "")
	arguments, _ := action["arguments"].(map[string]any)
	if arguments == nil {
		arguments = map[string]any{}
	}

	toolID, toolErr := registryFirst("tool_registry", "svc:tools")
	var result map[string]any
	if toolErr == nil && toolID != "" {
		// Use "execute" as msgType so SchemaValidationFacet (keyed by method name)
		// does not apply tool-specific schemas to the wrapper envelope.
		resp, askErr := askActor(toolID, "execute", map[string]any{
			"op":    "execute",
			"name":  toolName,
			"input": arguments,
		}, 5000)
		if askErr == nil {
			result = resp
		}
	}
	if result == nil {
		result = map[string]any{"error": "tool_registry unavailable", "tool": toolName}
	}

	inputTokens := intVal(result, "input_tokens", 0)
	outputTokens := intVal(result, "output_tokens", 0)
	return loop.ToolCall(toolName, arguments, result, inputTokens, outputTokens, "")
}

func (a *AgentActor) exportTrajectory(traj plexspaces.AgentTrajectory) {
	key := "agent_trajectory:" + traj.TrajectoryID
	trajJSON, err := json.Marshal(traj)
	if err != nil {
		host.Warn(fmt.Sprintf("AgentActor: failed to marshal trajectory: %v", err))
		return
	}
	host.KV().Put(key, string(trajJSON))

	// Update per-agent index
	indexKey := "agent_trajectory_index:" + a.ActorID
	existing := func() []string { v, _ := host.KV().Get(indexKey); return parseStringSlice(v) }()
	existing = append(existing, traj.TrajectoryID)
	indexJSON, _ := json.Marshal(existing)
	host.KV().Put(indexKey, string(indexJSON))

	// Write trajectory tuple to TupleSpace for eval collection
	_ = host.TS().Write([]any{"trajectory", traj.EvalRunID, traj.TrajectoryID, traj.Outcome})

	// Store in trajectory store if available
	tsID, err := registryFirst("trajectory_store", "svc:trajectory_store")
	if err == nil && tsID != "" {
		trajMap := map[string]any{}
		if err2 := json.Unmarshal(trajJSON, &trajMap); err2 == nil {
			_, _ = askActor(tsID, "put", map[string]any{
				"op":         "put",
				"trajectory": trajMap,
			}, 3000)
		}
	}
}

func (a *AgentActor) queryExecutionTrace() string {
	indexKey := "trace_index:" + a.ActorID
	indexRaw, _ := host.KV().Get(indexKey)
	if indexRaw != "" {
		var traceIDs []string
		if err := json.Unmarshal([]byte(indexRaw), &traceIDs); err == nil && len(traceIDs) > 0 {
			raw, _ := host.KV().Get("trace:" + traceIDs[len(traceIDs)-1])
			if raw != "" {
				var trace map[string]any
				if err2 := json.Unmarshal([]byte(raw), &trace); err2 == nil {
					return marshal(trace)
				}
			}
		}
	}
	return marshal(map[string]any{"actor_id": a.ActorID, "steps": []any{}, "outcome": "running"})
}

func (a *AgentActor) queryStatus() string {
	task := a.Task
	if len(task) > 80 {
		task = task[:80]
	}
	return marshal(map[string]any{
		"actor_id":        a.ActorID,
		"task":            task,
		"iterations_done": a.IterationsDone,
		"total_tool_calls": a.TotalToolCalls,
	})
}

File: examples/python/apps/minipi/agent.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""AgentActor — OODA-loop agent with durable execution trace capture.

Demonstrates the harness layer: loop control, tool calling, token budget,
crash recovery via DurabilityFacet, and execution trace export via
ExecutionTraceFacet.

Harness = agent - model (everything except the LLM).
"""

import json
from plexspaces import workflow_actor, state, init_handler, run_handler, signal_handler, query_handler, host
from plexspaces.agent import AgentLoop, AgentConfig

_MAX_ITER = 10
_TOKEN_BUDGET = 4096


@workflow_actor
class AgentActor:
    """
    OODA-loop agent: Observe → Orient → Decide → Act.

    Durable: DurabilityFacet journals every step. Crash at step 7?
    Restart brings back all prior steps from journal — no re-burned tokens.

    Execution Trace: ExecutionTraceFacet captures ordered step sequence for eval,
    writes trace:{id} and trace_index:{actor_id} to KV on completion.
    """

    actor_id: str = state(default="")
    task: str = state(default="")
    iterations_done: int = state(default=0)
    total_tool_calls: int = state(default=0)
    eval_run_id: str = state(default="")
    scenario_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        args = config.get("args", {})
        self.eval_run_id = args.get("eval_run_id", "")
        self.scenario_id = args.get("scenario_id", "")
        try:
            host.kv.put("svc:agent_runner", host.self_id())
        except Exception:
            pass
        try:
            host.registry.register(None, self.actor_id or host.self_id(), "actor", "",
                                   object_category="agent_runner")
        except Exception:
            pass
        host.info(f"AgentActor init actor_id={self.actor_id} eval_run={self.eval_run_id}")

    @run_handler
    def run(self, task: str = "", eval_run_id: str = "", scenario_id: str = "") -> dict:
        """Main OODA loop — durable: each step is journaled."""
        # SDK may pass the full payload dict as `task` when run_fn == instance.run
        if isinstance(task, dict):
            payload = task
            task = payload.get("task", "")
            eval_run_id = eval_run_id or payload.get("eval_run_id", "")
            scenario_id = scenario_id or payload.get("scenario_id", "")
        if not task:
            return {"error": "task is required"}

        self.task = task
        if eval_run_id:
            self.eval_run_id = eval_run_id
        if scenario_id:
            self.scenario_id = scenario_id

        host.info(f"AgentActor starting task: {str(task)[:80]}")

        loop = AgentLoop(
            actor_id=self.actor_id or host.actor_id(),
            config=AgentConfig(max_iterations=_MAX_ITER, token_budget=_TOKEN_BUDGET),
            eval_run_id=self.eval_run_id,
            scenario_id=self.scenario_id,
        )

        # OODA loop — runs until done, budget exceeded, or suspended
        while not loop.iteration_limit_reached():
            if loop.budget_exceeded():
                host.incr_counter("agent_budget_exceeded", 1)
                traj = loop.finalize_trajectory("budget_exceeded", f"Token budget {_TOKEN_BUDGET} exceeded")
                return {"status": "budget_exceeded", "trajectory": traj}

            if loop.is_suspended:
                traj = loop.get_trajectory().to_dict()
                return {"status": "suspended", "trajectory": traj}

            # --- OBSERVE: fetch context, memory, environment ---
            observations = self._do_observe(loop, task)

            # --- ORIENT: analyze observations with LLM ---
            plan = self._do_orient(loop, observations)

            # --- DECIDE: pick next action ---
            action = self._do_decide(loop, plan)

            if action.get("done"):
                break

            # Check for approval-required actions (human-in-the-loop)
            if action.get("needs_approval"):
                loop.suspend(f"action_needs_approval:{action.get('tool_name', 'unknown')}")
                traj = loop.get_trajectory().to_dict()
                return {"status": "suspended", "trajectory": traj}

            # --- ACT: execute the chosen tool ---
            result = self._do_act(loop, action)
            self.total_tool_calls += 1
            self.iterations_done += 1

            host.incr_counter("agent_iterations", 1)
            loop.increment_iteration()

        traj = loop.finalize_trajectory("completed", f"Completed {self.iterations_done} iterations")
        self._export_trajectory(traj)
        host.incr_counter("agent_runs_completed", 1)
        return {"status": "success", "task": task, "iterations": self.iterations_done, "trajectory": traj}

    @signal_handler("resume")
    def on_resume(self, data: dict = None) -> None:
        """Resume after human-in-the-loop approval."""
        host.info(f"AgentActor resumed: {data}")

    @query_handler("execution_trace")
    def on_query_execution_trace(self) -> dict:
        """Live execution trace inspection mid-run (returns most recent KV trace)."""
        try:
            index_raw = host.kv.get(f"trace_index:{self.actor_id}")
            if index_raw:
                trace_ids = json.loads(index_raw)
                if trace_ids:
                    raw = host.kv.get(f"trace:{trace_ids[-1]}")
                    if raw:
                        return json.loads(raw)
        except Exception:
            pass
        return {"actor_id": self.actor_id, "steps": [], "outcome": "running"}

    @query_handler("status")
    def on_query_status(self) -> dict:
        return {
            "actor_id": self.actor_id,
            "task": self.task[:80] if self.task else "",
            "iterations_done": self.iterations_done,
            "total_tool_calls": self.total_tool_calls,
        }

    # ------------------------------------------------------------------
    # OODA implementation helpers

    def _do_observe(self, loop: AgentLoop, task: str) -> dict:
        """OBSERVE: gather context from memory and environment."""
        memory_key = f"agent_memory:{self.actor_id}"
        prior_context = {}
        try:
            raw = host.kv.get(memory_key)
            if raw:
                prior_context = json.loads(raw)
        except Exception:
            pass

        observations = {
            "task": task,
            "prior_context": prior_context,
            "iteration": self.iterations_done,
        }
        return loop.observe(observations)

    def _do_orient(self, loop: AgentLoop, observations: dict) -> dict:
        """ORIENT: analyze observations with LLM to build a plan."""
        llm_id = self._find_service("llm_gateway")
        if not llm_id:
            plan = {
                "analysis": f"Processing task: {observations.get('task', '')}",
                "next_tool": "calculator",
                "arguments": {"expression": observations.get('task', '1+1')},
                "done": False,
            }
        else:
            messages = [
                {"role": "system", "content": "You are a helpful agent. Analyze the task and decide what to do next."},
                {"role": "user", "content": f"Task: {observations.get('task', '')}\nIteration: {observations.get('iteration', 0)}"},
            ]
            resp = self._ask(llm_id, "completion", {"messages": messages}, timeout_ms=10000)
            if not resp or "error" in resp:
                plan = {"done": True, "result": "LLM unavailable"}
            else:
                response = resp.get("response", {})
                plan = {
                    "analysis": response.get("content", ""),
                    "next_tool": response.get("tool_name", "calculator"),
                    "arguments": response.get("arguments", {}),
                    "input_tokens": resp.get("input_tokens", 0),
                    "output_tokens": resp.get("output_tokens", 0),
                    "model": resp.get("model", ""),
                    "done": response.get("stop_reason") == "end_turn" and not response.get("tool_calls"),
                }

        return loop.orient(plan)

    def _do_decide(self, loop: AgentLoop, plan: dict) -> dict:
        """DECIDE: select the next concrete action."""
        action = {
            "tool_name": plan.get("next_tool", "calculator"),
            "arguments": plan.get("arguments", {}),
            "done": plan.get("done", False),
            "needs_approval": plan.get("needs_approval", False),
        }
        return loop.decide(action)

    def _do_act(self, loop: AgentLoop, action: dict) -> dict:
        """ACT: execute the tool call."""
        tool_name = action.get("tool_name", "")
        arguments = action.get("arguments", {})

        tool_id = self._find_service("tool_registry")
        if not tool_id:
            result = {"error": "tool_registry unavailable", "tool": tool_name}
        else:
            result = self._ask(tool_id, tool_name, arguments) or {}

        return loop.tool_call(
            name=tool_name,
            args=arguments,
            result=result,
            input_tokens=result.get("input_tokens", 0),
            output_tokens=result.get("output_tokens", 0),
        )

    def _export_trajectory(self, traj: dict) -> None:
        """Store completed trajectory in KV (ExecutionTraceFacet also exports its own trace)."""
        try:
            key = f"agent_trajectory:{traj.get('trajectory_id', '')}"
            host.kv.put(key, json.dumps(traj))

            # Update per-agent index for easy retrieval by eval runner
            index_key = f"agent_trajectory_index:{self.actor_id}"
            try:
                existing = json.loads(host.kv.get(index_key) or "[]")
            except Exception:
                existing = []
            existing.append(traj.get("trajectory_id", ""))
            host.kv.put(index_key, json.dumps(existing))
        except Exception as e:
            host.warn(f"Failed to export trajectory: {e}")

    def _find_service(self, service_type: str) -> str:
        """Discover service actor ID via object registry; falls back to peer ID on same node."""
        try:
            regs = host.registry.discover(None, object_category=service_type, limit=1)
            if regs:
                return regs[0]["object_id"]
        except Exception:
            pass
        idx = self.actor_id.find("//")
        if idx >= 0:
            return service_type + self.actor_id[idx:]
        return service_type

    def _ask(self, actor_id: str, op: str, payload: dict, timeout_ms: int = 5000) -> dict:
        """Send a request-reply message to another actor."""
        try:
            result = host.ask(actor_id, op, payload, timeout_ms)
            return result or {}
        except Exception as e:
            return {"error": str(e)}
